packages/common/src/search-attributes.ts

```diff
@@ -0,0 +1,300 @@
+  private readonly _type: T;
+  private readonly _value: V;
+
+  constructor(type: T, value: V) {
+    this._type = type;
+    this._value = value;
+  }
+
+  get type(): T {
+    return this._type;
+  }
+
+  get value(): V {
+    return this._value;
+  }
+}
+
+// Internal type for class private data.
+// Exported for use in payload conversion.
+export class TypedSearchAttributeValue<T extends SearchAttributeType> extends BaseSearchAttributeValue<T> {}
+// ts-prune-ignore-next
+export class TypedSearchAttributeUpdateValue<T extends SearchAttributeType> extends BaseSearchAttributeValue<
+  T,
+  IndexedValueTypeMapping[T] | null
+> {}
+
+export type SearchAttributePair = {
+  [T in SearchAttributeType]: { key: SearchAttributeKey<T>; value: IndexedValueTypeMapping[T] };
+}[SearchAttributeType];
+
+export type SearchAttributeUpdatePair = {
+  [T in SearchAttributeType]: { key: SearchAttributeKey<T>; value: IndexedValueTypeMapping[T] | null };
+}[SearchAttributeType];
+
+export class TypedSearchAttributes {
+  private searchAttributes: Record<string, TypedSearchAttributeValue<SearchAttributeType>> = {};
+
+  constructor(initialAttributes?: SearchAttributePair[]) {
+    if (initialAttributes === undefined) return;
+    for (const pair of initialAttributes) {
+      if (pair.key.name in this.searchAttributes) {
+        throw new Error(`Duplicate search attribute key: ${pair.key.name}`);
+      }
+      this.searchAttributes[pair.key.name] = new TypedSearchAttributeValue(pair.key.type, pair.value);
+    }
+  }
+
+  get<T extends SearchAttributeType>(key: SearchAttributeKey<T>): IndexedValueTypeMapping[T] | undefined {
+    const attr = this.searchAttributes[key.name];
+    // Key not found or type mismatch.
+    if (attr === undefined || !isValidValueForType(key.type, attr.value)) {
+      return undefined;
+    }
+    return attr.value;
+  }
+
+  /** Returns a deep copy of the given TypedSearchAttributes instance */
+  copy(): TypedSearchAttributes {
+    const state: Record<string, TypedSearchAttributeValue<SearchAttributeType>> = {};
+
+    for (const [key, attr] of Object.entries(this.searchAttributes)) {
+      // Create a new instance with the same properties
+      let value = attr.value;
+      // For non-primitive types, create a deep copy
+      if (attr.value instanceof Date) {
+        value = new Date(attr.value);
+      } else if (Array.isArray(attr.value)) {
+        value = [...attr.value];
+      }
+      state[key] = new TypedSearchAttributeValue(attr.type, value);
+    }
+
+    // Create return value with manually assigned state.
+    const res = new TypedSearchAttributes();
+    res.searchAttributes = state;
+    return res;
+  }
+
+  /**
+   * Return JSON representation of this class as SearchAttributePair[]
+   * Default toJSON method is not used because it's JSON representation includes private state.
+   */
+  toJSON(): SearchAttributePair[] {
```

Comment:
Can add a `@hidden` annotation on this method, we don't need it to appear in API docs.

Reply:
Done.